### PR TITLE
Lampo updated to LDK 0.0.123

### PR DESCRIPTION
**Pull request by Psycho-Pirate**
Updated LDK to 0.0.123
Updated bdk to 1.0.0-alpha.11

Commits:
* meta: bdk upgrade (adding in the body that this change is required by the bdk change)
  *  Updated lampo-core-wallet/Cargo.toml with latest bdk version
  *  Used `Xpriv` instead of `ExtendedPrivKey` in lampo-core-wallet/src/lib.rs
* meta: ldk upgrade
  * Used `Option<PublicKey>` instead of `String` in lampo-common/src/model/invoice.rs
  * Added support for `ldk::events::PaymentPurpose::Bolt12OfferPayment` and `ldk::events::PaymentPurpose::Bolt12RefundPayment` in lampod/src/actions/handler.rs
  * Added support for `ConfirmationTarget::OutputSpendingFee` in lampod/src/chain/blockchain.rs
  * Updated `create_offer_builder()` in lampod/src/jsonrpc/offchain.rs
  * Updated `destination` in lampod/src/ln/onion_message.rs
  * Updated `OnionMessenger` in lampod/src/ln/peer_manager.rs

### CODE REVIEW

**Comment by vincenzopalazzo** [comment]
```
Run actions-rs/cargo@v1
/home/runner/.cargo/bin/cargo fmt --all -- --check
Diff in /home/runner/work/lampo.rs/lampo.rs/lampo-common/src/model/invoice.rs at line 54:
             Self {
                 bolt12: value.to_string(),
                 metadata: value.metadata().map(|bytes| hex::encode(bytes)),
-                metadata_pubkey: value.signing_pubkey().expect("Public key is missing").to_string(),
+                metadata_pubkey: value
+                    .signing_pubkey()
+                    .expect("Public key is missing")
+                    .to_string(),
             }
         }
     }
Diff in /home/runner/work/lampo.rs/lampo.rs/lampod/src/chain/blockchain.rs at line 4[6](https://github.com/vincenzopalazzo/lampo.rs/actions/runs/9330760464/job/25684820643?pr=240#step:4:7):
             ConfirmationTarget::MinAllowedNonAnchorChannelRemoteFee => {
                 String::from("min_allowed_non_anchor_channel_remote")
             }
-            ConfirmationTarget::OutputSpendingFee => {
-                String::from("output_spending")
-            }
+            ConfirmationTarget::OutputSpendingFee => String::from("output_spending"),
         }
     }
 
Error: The process '/home/runner/.cargo/bin/cargo' failed with exit code 1
```

File: lampo-common/src/model/invoice.rs [hunk 1]
```diff
@@ -54,7 +54,10 @@ pub mod response {
             Self {
                 bolt12: value.to_string(),
                 metadata: value.metadata().map(|bytes| hex::encode(bytes)),
-                metadata_pubkey: value.signing_pubkey().to_string(),
+                metadata_pubkey: value
+                    .signing_pubkey()
+                    .expect("Public key is missing")
```
Comment (vincenzopalazzo) [inline lampo-common/src/model/invoice.rs:59]: We do not use `expect` inside lampo, or we add a safety comment (look at some example with `git grep -n SAFETY`)

Expect is just an wrong suggestion of clippy

File: lampo-core-wallet/src/lib.rs [hunk 2]
```diff
@@ -72,7 +72,7 @@ impl CoreWalletManager {
         xprv: lampo_common::bitcoin::PrivateKey,
         channel_keys: Option<String>,
     ) -> error::Result<(bdk::Wallet, LampoKeys)> {
-        use bdk::bitcoin::bip32::ExtendedPrivKey;
+        use bdk::bitcoin::bip32::Xpriv;
```
Comment (vincenzopalazzo) [inline lampo-core-wallet/src/lib.rs:75]: mh this should be added in a separate commit, like adding a commit to upgrade bdk and then add another commit to upgrade ldk

File: lampod/src/actions/handler.rs [hunk 3]
```diff
@@ -263,12 +264,13 @@ impl Handler for LampoHandler {
...
             } => {
                 let (payment_preimage, payment_secret) = match purpose {
-                    ldk::events::PaymentPurpose::InvoicePayment {
+                    ldk::events::PaymentPurpose::Bolt11InvoicePayment {
                         payment_preimage,
                         payment_secret,
                         ..
                     } => (payment_preimage, Some(payment_secret)),
                     ldk::events::PaymentPurpose::SpontaneousPayment(preimage) => (Some(preimage), None),
+                    _ => panic!("BOLT 12 not implemented"),
```
Comment (vincenzopalazzo) [inline lampod/src/actions/handler.rs:273]: This is wrong, lampo support bolt12, so we should look inside what it is going on here

File: lampod/src/chain/blockchain.rs [hunk 4]
```diff
@@ -87,6 +89,9 @@ impl FeeEstimator for LampoChainManager {
             ConfirmationTarget::ChannelCloseMinimum => {
                 self.backend.fee_rate_estimation(100).unwrap_or_default()
             }
+            ConfirmationTarget::OutputSpendingFee => {
+                self.backend.fee_rate_estimation(1).unwrap_or_default()
```
Comment (vincenzopalazzo) [inline lampod/src/chain/blockchain.rs:93]: why 1 block of estimation fee? there is any reason? what the ldk docs says?

File: lampod/src/jsonrpc/offchain.rs [hunk 5]
```diff
@@ -48,7 +48,7 @@ pub fn json_offer(ctx: &LampoDaemon, request: &json::Value) -> Result<json::Valu
     let request: GenerateOffer = json::from_value(request.clone())?;
     let manager = ctx.channel_manager().manager();
     let mut offer_builder = manager
-        .create_offer_builder(request.description)
```
Comment (vincenzopalazzo) [inline lampod/src/jsonrpc/offchain.rs:51]: we should see how it is possible to add the description inside the offer. You are adding a regression in this case

File: lampod/src/ln/onion_message.rs [hunk 6]
```diff
@@ -96,9 +97,10 @@ where
         let first_node = match destination {
             Destination::Node(node_id) => node_id,
             Destination::BlindedPath(BlindedPath {
-                introduction_node_id: node_id,
+                introduction_node: IntroductionNode::NodeId(node_id),
                 ..
             }) => node_id,
+            _ => panic!("Invalid destination"),
```
Comment (vincenzopalazzo) [inline lampod/src/ln/onion_message.rs:103]: we should return the error, lampo never panic

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]
Ok the commit need to be squash in a single one because the second one is not useful.

File: lampo-core-wallet/src/lib.rs (hunk 2 above)
Comment (Psycho-Pirate) [reply lampo-core-wallet/src/lib.rs:75]: Okay

File: lampo-common/src/model/invoice.rs (hunk 1 above)
Comment (Psycho-Pirate) [reply lampo-common/src/model/invoice.rs:59]: So should I simply unwrap this and add a SAFETY comment?

File: lampod/src/ln/onion_message.rs (hunk 6 above)
Comment (Psycho-Pirate) [reply lampod/src/ln/onion_message.rs:103]:
```
diff --git a/lampod/src/ln/onion_message.rs b/lampod/src/ln/onion_message.rs
index 3ce49c6..dcf553c 100644
--- a/lampod/src/ln/onion_message.rs
+++ b/lampod/src/ln/onion_message.rs
@@ -95,13 +95,13 @@ where
         destination: lampo_common::ldk::onion_message::messenger::Destination,
     ) -> Result<lampo_common::ldk::onion_message::messenger::OnionMessagePath, ()> {
         let first_node = match destination {
-            Destination::Node(node_id) => node_id,
+            Destination::Node(node_id) => Ok(node_id),
             Destination::BlindedPath(BlindedPath {
                 introduction_node: IntroductionNode::NodeId(node_id),
                 ..
-            }) => node_id,
-            _ => panic!("Invalid destination"),
-        };
+            }) => Ok(node_id),
+            _ => Err("Invalid destination".to_string()),
+        }.unwrap();
```

Is this okay?

Comment (vincenzopalazzo) [reply lampod/src/ln/onion_message.rs:103]: I do not know, maybe we should support the destination, why fo you it is an invalid destination?

File: lampod/src/jsonrpc/offchain.rs (hunk 5 above)
Comment (Psycho-Pirate) [reply lampod/src/jsonrpc/offchain.rs:51]: They have removed the description and they are sending none as description.
https://docs.rs/lightning/0.0.122/src/lightning/offers/offer.rs.html#186-203
https://docs.rs/lightning/latest/src/lightning/offers/offer.rs.html#257-274

Comment (vincenzopalazzo) [reply lampod/src/jsonrpc/offchain.rs:51]: untrue, https://docs.rs/lightning/0.0.122/src/lightning/offers/offer.rs.html#1017

Description is inside the BOLT12 spec anyway

File: lampod/src/ln/onion_message.rs [hunk 7]
```diff
@@ -94,13 +94,14 @@ where
...
             Destination::Node(node_id) => node_id,
             Destination::BlindedPath(BlindedPath {
-                introduction_node: IntroductionNode::NodeId(node_id),
-                ..
-            }) => node_id,
-            _ => panic!("Invalid destination"),
+                introduction_node, ..
+            }) => match introduction_node {
+                IntroductionNode::NodeId(node_id) => node_id,
+                IntroductionNode::DirectedShortChannelId(..) => return Err(()),
```
Comment (vincenzopalazzo) [inline lampod/src/ln/onion_message.rs:103]: not sure that this `return Err(())` is right but anyway it is fine for now

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]
Ok running the CI on it, but please now organize the commit in a better way.

Something like this

- meta: bdk upgrade (adding in the body that this change is required by the bdk change)
- meta: ldk upgrade

**Review by vincenzopalazzo — APPROVED** [review]
LGTM
